yt/visualization/_handlers.py

```diff
@@ -0,0 +1,361 @@
+    dropped, and vice versa. These rules are implemented with properties and
+    watcher patterns.
+
+    It also keeps track of display units so that vmin, vmax and linthresh can be
+    updated with implicit units.
+    """
+
+    # using slots here to minimize the risk of introducing bugs
+    # since attributes names are essential to this class's implementation
+    __slots__ = (
+        "data_source",
+        "ds",
+        "_display_units",
+        "_vmin",
+        "_vmax",
+        "_norm_type",
+        "_linthresh",
+        "_norm_type",
+        "_norm",
+    )
+    _constraint_attrs: List[str] = ["vmin", "vmax", "norm_type", "linthresh"]
+
+    def __init__(
+        self,
+        data_source,
+        *,
+        display_units: Unit,
+        vmin: Optional[unyt_quantity] = None,
+        vmax: Optional[unyt_quantity] = None,
+        norm_type: Optional[Type[Normalize]] = None,
+        norm: Optional[Normalize] = None,
+        linthresh: Optional[float] = None,
+    ):
+        self.data_source = weakref.proxy(data_source)
+        self.ds = data_source.ds  # should already be a weakref proxy
+        self._display_units = display_units
+
+        self._norm = norm
+        self._vmin = vmin
+        self._vmax = vmax
+        self._norm_type = norm_type
+        self._linthresh = linthresh
+
+        if self.has_norm and self.has_constraints:
+            raise TypeError(
+                "NormHandler input is malformed. "
+                "A norm cannot be passed along other constraints."
+            )
+
+    def _get_constraints(self) -> Dict[str, Any]:
+        return {
+            attr: getattr(self, attr)
+            for attr in self.__class__._constraint_attrs
+            if getattr(self, attr) is not None
+        }
+
+    @property
+    def has_constraints(self) -> bool:
+        return bool(self._get_constraints())
+
+    def _reset_constraints(self) -> None:
+        constraints = self._get_constraints()
+        if not constraints:
+            return
+
+        msg = ", ".join([f"{name}={value}" for name, value in constraints.items()])
+        mylog.warning("Dropping norm constraints (%s)", msg)
+        for name in constraints.keys():
+            setattr(self, name, None)
+
+    @property
+    def has_norm(self) -> bool:
+        return self._norm is not None
+
+    def _reset_norm(self):
+        if not self.has_norm:
+            return
+        mylog.warning("Dropping norm (%s)", self.norm)
+        self._norm = None
+
+    def to_float(self, val: unyt_quantity) -> float:
+        return float(val.to(self.display_units).d)
+
+    def to_quan(self, val) -> unyt_quantity:
+        if isinstance(val, unyt_quantity):
+            return self.ds.quan(val)
+        elif (
+            is_sequence(val)
+            and len(val) == 2
+            and isinstance(val[0], Real)
+            and isinstance(val[1], (str, Unit))
+        ):
+            return self.ds.quan(*val)
+        elif isinstance(val, Real):
```

Comment:
`float` is a special case in Python typing: type checkers are perfectly happy with this:
```python
def foo(x: float):
   pass
foo(1)
```
this is because in Python, anywhere a `float` is expected, an `int` also works, as far as I understand, *except* that  `int` doesn't inherit from float, so `isinstance(1, float)` evaluates to `False`, which is why we need something like `Real` for runtime type checking. Also note that numpy integers and floats types are compatible with `Real`.